compiler/syntax.go

```diff
@@ -379,6 +379,20 @@ func (c *syntaxLoader) convertSet(expr ast.SetExpression, nonterm *syntax.Nonter
 			default:
 				c.Errorf(op, "operator must be one of: first, last, precede or follow")
 			}
+		} else {
+			// Check if the reference targets another named set.
+			name := expr.Symbol().Name()
+			if index, ok := c.namedSets[name.Text()]; ok {
+				if index >= len(c.out.Sets) {
+					c.Errorf(expr.TmNode(), "forward references are not allowed")
+					return ret
+				}
+				if _, ok := c.resolveRefIndex(expr.Symbol()); ok {
+					c.Errorf(expr.TmNode(), "ambigious reference")
+					return ret
+				}
+				return c.out.Sets[index]
```

Comment:
This works but what happens here under the hood is that the set expression you reference here is evaluated several times. I'd instead introduce a new syntax.SetOp - Reference, and store the SetIndex inside TokenSet (similar to syntax.Expr.SetIndex). Then inside syntax/set->translate function create a future set.